### DISCUSSION TITLE
UseCaseKey usage

### DISCUSSION THREAD

**Discussion by maria-sanchez-gs** [archimedes-projects/archimedes-js#38] [Q&A]
I am using `@UseCaseKey` annotation in some Qry and Cmd in order to Invalidate Archimedes Cache for the annotated use cases.

I wonder if it might be the case of **wanting or needing to annotate a qry/cmd with a string other than its class name**.

I.e.
```ts
@UseCaseKey('FooQry')
@Injectable()
export class FooQry extends Query<Result, Params> {
...
}
```

**Comment by cesalberca** [comment]
Hello María! Thanks for the question. 

First some context: we use the `UseCaseKey` decorator since some bundlers when they create the bundle and minimise the code they might change the class names from `FooQry` to `_a` and things like that. Sometimes they use the same name for different classes making the cache invalidate when they shouldn't. The ideal would be to use `constructor.name`, but since we can't rely on the bundler we added support for this decorator. In the future maybe it won't be needed once decorators land on JavaScript.

Now, answering your question: another use case (pun intended) would be to define different scopes. For example you could create a `User` scope and invalidate the whole scope when needed.

```
@UseCaseKey('User')
@Injectable()
export class ProfileQry extends Query<Result, Params> {
}
```


```
@UseCaseKey('User')
@Injectable()
export class RoleQry extends Query<Result, Params> {
}
```

```
@UseCaseKey('ChangeUserCmd')
@Injectable()
export class ChangeUserCmd extends Query<Result, Params> {
}
```

```
CacheInvalidations.set('ChangeUserCmd', ['User'])
```

**Reply by maria-sanchez-gs** [reply]
That's an interesting use of the decorator.

Thanks for the clarification and context, @cesalberca. Keep up the good work!
